File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/weaveit/seatingplacesuggestions/SeatingArrangementRecommender.java
package org.weaveit.seatingplacesuggestions;

import java.util.ArrayList;
import java.util.List;

public class SeatingArrangementRecommender {
    private static final int NUMBER_OF_SUGGESTIONS = 3;
    private final AuditoriumSeatingArrangements auditoriumSeatingArrangements;

    public SeatingArrangementRecommender(AuditoriumSeatingArrangements auditoriumSeatingArrangements) {
        this.auditoriumSeatingArrangements = auditoriumSeatingArrangements;
    }

    public SuggestionsAreMade makeSuggestions(String showId, int partyRequested) {
        var auditoriumSeating = auditoriumSeatingArrangements.findByShowId(showId);

        var allSuggestions = new ArrayList<SuggestionIsMade>();
        allSuggestions.addAll(giveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.FIRST));
        allSuggestions.addAll(giveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.SECOND));
        allSuggestions.addAll(giveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.THIRD));
        allSuggestions.addAll(giveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.MIXED));

        var suggestionsMade = new SuggestionsAreMade(showId, partyRequested, allSuggestions);

        if (suggestionsMade.matchExpectations())
            return suggestionsMade;

        return new SuggestionsAreNotAvailable(showId, partyRequested);
    }

    private static List<SuggestionIsMade> giveMeSuggestionsFor(
            AuditoriumSeatingArrangement auditoriumSeatingArrangement, int partyRequested, PricingCategory pricingCategory) {
        var foundSuggestions = new ArrayList<SuggestionIsMade>();
        var currentArrangement = auditoriumSeatingArrangement;

        for (int i = 0; i < NUMBER_OF_SUGGESTIONS; i++) {
            var seatingOptionSuggested = currentArrangement.suggestSeatingOptionFor(partyRequested, pricingCategory);

            if (seatingOptionSuggested.matchExpectation()) {
                currentArrangement = currentArrangement.allocate(seatingOptionSuggested.seats());
                foundSuggestions.add(new SuggestionIsMade(seatingOptionSuggested));
            }
        }

        return foundSuggestions;
    }
}

File: TheaterSuggestions/Kotlin/ScreeningSeatingArrangementSuggestions/src/main/kotlin/org/weaveit/seatingplacesuggestions/SeatingArrangementRecommender.kt
package org.weaveit.seatingplacesuggestions

class SeatingArrangementRecommender(
    private val auditoriumSeatingArrangements: AuditoriumSeatingArrangements
) {
    companion object {
        private const val NUMBER_OF_SUGGESTIONS = 3
    }

    fun makeSuggestions(showId: String, partyRequested: Int): SuggestionsAreMade {
        val auditoriumSeating = auditoriumSeatingArrangements.findByShowId(showId)

        val suggestionsMade = SuggestionsAreMade(showId, partyRequested)

        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.FIRST))
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.SECOND))
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.THIRD))

        if (suggestionsMade.matchExpectations()) {
            return suggestionsMade
        }

        return SuggestionsAreNotAvailable(showId, partyRequested)
    }

    private fun giveMeSuggestionsFor(
        auditoriumSeatingArrangement: AuditoriumSeatingArrangement,
        partyRequested: Int,
        pricingCategory: PricingCategory
    ): List<SuggestionIsMade> {
        val foundedSuggestions = mutableListOf<SuggestionIsMade>()

        for (i in 0 until NUMBER_OF_SUGGESTIONS) {
            val seatingOptionSuggested = auditoriumSeatingArrangement.suggestSeatingOptionFor(partyRequested, pricingCategory)

            if (seatingOptionSuggested.matchExpectation()) {
                for (seatingPlace in seatingOptionSuggested.seats()) {
                    seatingPlace.allocate()
                }

                foundedSuggestions.add(SuggestionIsMade(seatingOptionSuggested))
            }
        }

        return foundedSuggestions
    }
}
